File: python/src/webkaya/policy.py
"""Sandbox governance policy (Python port of ``src/runtime/policy.ts``)."""

from __future__ import annotations

from dataclasses import dataclass, replace
from typing import List, Optional

MAX_POLICY_CODE_LENGTH = 20_000

# The guest sees no ambient I/O: it interacts only through ``ctx``. These tokens
# are rejected to keep runs local, replayable, and free of hidden dependencies.
DISALLOWED_GUEST_TOKENS = (
    "import ",
    "__",
    "open(",
    "eval(",
    "exec(",
    "compile(",
    "globals(",
    "locals(",
    "getattr(",
    "setattr(",
    "delattr(",
    "input(",
    "breakpoint(",
    "vars(",
)


@dataclass(frozen=True)
class SandboxPolicy:
    entrypoint: str = "main"
    timeout_ms: int = 10_000
    retry_count: int = 1
    memory_budget_mb: int = 512
    cold_start_ms: int = 0
    max_guest_code_length: int = 20_000
    snapshot_every_n_runs: int = 0


DEFAULT_SANDBOX_POLICY = SandboxPolicy()


def _clamp(value: int, low: int, high: int) -> int:
    return min(high, max(low, value))


def _to_int(value, fallback: int) -> int:
    try:
        return int(value)
    except (TypeError, ValueError):
        return fallback


def _sanitize_entrypoint(raw) -> str:
    if not isinstance(raw, str):
        return DEFAULT_SANDBOX_POLICY.entrypoint
    cleaned = "".join(ch if ch.isalnum() or ch == "_" else "_" for ch in raw.strip())
    return cleaned or DEFAULT_SANDBOX_POLICY.entrypoint


def normalize_policy(policy: Optional[dict] = None) -> SandboxPolicy:
    p = dict(policy or {})
    return SandboxPolicy(
        entrypoint=_sanitize_entrypoint(p.get("entrypoint", DEFAULT_SANDBOX_POLICY.entrypoint)),
        timeout_ms=_clamp(_to_int(p.get("timeout_ms", 10_000), 10_000), 100, 120_000),
        retry_count=_clamp(_to_int(p.get("retry_count", 1), 1), 0, 8),
        memory_budget_mb=_clamp(_to_int(p.get("memory_budget_mb", 512), 512), 64, 16_384),
        cold_start_ms=_clamp(_to_int(p.get("cold_start_ms", 0), 0), 0, 10_000),
        max_guest_code_length=_clamp(_to_int(p.get("max_guest_code_length", 20_000), 20_000), 256, 200_000),
        snapshot_every_n_runs=_clamp(_to_int(p.get("snapshot_every_n_runs", 0), 0), 0, 1_000),
    )


def validate_policy(policy: SandboxPolicy) -> List[str]:
    errors: List[str] = []
    if not policy.entrypoint.strip():
        errors.append("Entrypoint name is required.")
    if not 100 <= policy.timeout_ms <= 120_000:
        errors.append("Timeout must be between 100 and 120000 ms.")
    if not 0 <= policy.retry_count <= 8:
        errors.append("Retry count must be between 0 and 8.")
    if not 64 <= policy.memory_budget_mb <= 16_384:
        errors.append("Memory budget must be between 64MB and 16384MB.")
    if not 0 <= policy.cold_start_ms <= 10_000:
        errors.append("Cold start delay must be between 0 and 10000 ms.")
    if not 256 <= policy.max_guest_code_length <= 200_000:
        errors.append("Max guest code length must be between 256 and 200000 characters.")
    if not 0 <= policy.snapshot_every_n_runs <= 1_000:
        errors.append("Snapshot cadence must be between 0 and 1000 runs.")
    return errors


def assert_guest_code_safety(code: str, max_length: int = MAX_POLICY_CODE_LENGTH) -> None:
    if len(code) > max_length:
        raise ValueError(f"Guest code too large (max {max_length} characters).")
    for token in DISALLOWED_GUEST_TOKENS:
        if token in code:
            raise ValueError(f'Guest code uses disallowed token: "{token.strip()}".')


__all__ = [
    "SandboxPolicy",
    "DEFAULT_SANDBOX_POLICY",
    "DISALLOWED_GUEST_TOKENS",
    "normalize_policy",
    "validate_policy",
    "assert_guest_code_safety",
    "replace",
]

File: src/runtime/policy.ts
import { PolicyEditorState, SandboxPolicy } from '../types/policy.js';

export const MAX_POLICY_CODE_LENGTH = 20_000;

/**
 * Tokens disallowed in any guest-supplied source (policy code and run code).
 * The sandbox provides no ambient I/O: guests interact only through `ctx`.
 * This keeps runs local, replayable, and free of hidden network dependencies.
 */
export const DISALLOWED_GUEST_TOKENS = [
  'import ',
  'require(',
  'fetch(',
  'XMLHttpRequest',
  'WebSocket',
  'Worker(',
  'new Function',
  'eval(',
  'globalThis',
  'window.',
  'document.',
  'navigator.',
  'self.',
  'postMessage(',
];

export const DEFAULT_SANDBOX_POLICY: SandboxPolicy = {
  entrypoint: 'main',
  timeoutMs: 10_000,
  retryCount: 1,
  memoryBudgetMB: 512,
  coldStartMs: 0,
  maxGuestCodeLength: 20_000,
  snapshotEveryNRuns: 0,
};

function clamp(value: number, min: number, max: number): number {
  return Math.min(max, Math.max(min, value));
}

function toNumber(value: unknown, fallback: number): number {
  const num = typeof value === 'number' ? value : Number(value);
  return Number.isFinite(num) ? num : fallback;
}

function sanitizeEntrypoint(raw: unknown): string {
  if (typeof raw !== 'string') return DEFAULT_SANDBOX_POLICY.entrypoint;
  const cleaned = raw
    .trim()
    .replace(/\s+/g, '_')
    .replace(/[^a-zA-Z0-9_]/g, '_');
  return cleaned || DEFAULT_SANDBOX_POLICY.entrypoint;
}

export function normalizePolicy(policy?: Partial<SandboxPolicy> | null): SandboxPolicy {
  const merged = { ...DEFAULT_SANDBOX_POLICY, ...(policy ?? {}) };
  return {
    entrypoint: sanitizeEntrypoint(merged.entrypoint),
    timeoutMs: clamp(Math.round(toNumber(merged.timeoutMs, 10_000)), 100, 120_000),
    retryCount: clamp(Math.round(toNumber(merged.retryCount, 1)), 0, 8),
    memoryBudgetMB: clamp(Math.round(toNumber(merged.memoryBudgetMB, 512)), 64, 16_384),
    coldStartMs: clamp(Math.round(toNumber(merged.coldStartMs, 0)), 0, 10_000),
    maxGuestCodeLength: clamp(Math.round(toNumber(merged.maxGuestCodeLength, 20_000)), 256, 200_000),
    snapshotEveryNRuns: clamp(Math.round(toNumber(merged.snapshotEveryNRuns, 0)), 0, 1_000),
  };
}

export function validatePolicy(policy: SandboxPolicy): string[] {
  const errors: string[] = [];
  if (!policy.entrypoint.trim()) errors.push('Entrypoint name is required.');
  if (policy.timeoutMs < 100 || policy.timeoutMs > 120_000) {
    errors.push('Timeout must be between 100 and 120000 ms.');
  }
  if (policy.retryCount < 0 || policy.retryCount > 8) errors.push('Retry count must be between 0 and 8.');
  if (policy.memoryBudgetMB < 64 || policy.memoryBudgetMB > 16_384) {
    errors.push('Memory budget must be between 64MB and 16384MB.');
  }
  if (policy.coldStartMs < 0 || policy.coldStartMs > 10_000) {
    errors.push('Cold start delay must be between 0 and 10000 ms.');
  }
  if (policy.maxGuestCodeLength < 256 || policy.maxGuestCodeLength > 200_000) {
    errors.push('Max guest code length must be between 256 and 200000 characters.');
  }
  if (policy.snapshotEveryNRuns < 0 || policy.snapshotEveryNRuns > 1_000) {
    errors.push('Snapshot cadence must be between 0 and 1000 runs.');
  }
  return errors;
}

export function assertGuestCodeSafety(code: string, maxLength = MAX_POLICY_CODE_LENGTH): void {
  if (code.length > maxLength) {
    throw new Error(`Guest code too large (max ${maxLength} characters).`);
  }
  for (const token of DISALLOWED_GUEST_TOKENS) {
    if (code.includes(token)) {
      throw new Error(`Guest code uses disallowed token: "${token.trim()}".`);
    }
  }
}

export function generatePolicyCodeFromTemplate(template: Partial<SandboxPolicy>): string {
  const cfg = normalizePolicy(template);
  return `// WebKaya Sandbox — ejected policy
// "api.base" is the safe baseline from template mode.
// Return a partial policy object to override fields.

const policy = {
  entrypoint: ${JSON.stringify(cfg.entrypoint)},
  timeoutMs: ${cfg.timeoutMs},
  retryCount: ${cfg.retryCount},
  memoryBudgetMB: ${cfg.memoryBudgetMB},
  coldStartMs: ${cfg.coldStartMs},
  maxGuestCodeLength: ${cfg.maxGuestCodeLength},
  snapshotEveryNRuns: ${cfg.snapshotEveryNRuns},
};

// Example tweak:
// policy.timeoutMs = api.int(policy.timeoutMs + 500, 100, 120000);

return policy;
`;
}

export function normalizePolicyEditorState(editor?: Partial<PolicyEditorState> | null): PolicyEditorState {
  const mode = editor?.mode === 'eject' ? 'eject' : 'template';
  const template = normalizePolicy(editor?.template);
  const code =
    typeof editor?.executableCode === 'string' && editor.executableCode.trim()
      ? editor.executableCode
      : generatePolicyCodeFromTemplate(template);
  return { mode, template, executableCode: code };
}

export function compilePolicyCode(code: string, basePolicy: Partial<SandboxPolicy>): SandboxPolicy {
  const trimmed = code.trim();
  if (!trimmed) {
    throw new Error('Ejected policy code is empty. Provide code or switch back to template mode.');
  }
  assertGuestCodeSafety(trimmed);
  const base = normalizePolicy(basePolicy);
  const api = {
    base: { ...base },
    clamp: (value: number, min: number, max: number) => clamp(value, min, max),
    int: (value: number, min: number, max: number) => clamp(Math.round(value), min, max),
    num: (value: number, min: number, max: number) => clamp(value, min, max),
  };

  let result: unknown;
  try {
    // eslint-disable-next-line no-new-func
    const executable = new Function('api', `"use strict";\n${trimmed}`);
    result = executable(api);
  } catch (error) {
    const message = error instanceof Error ? error.message : 'Unknown compile error.';
    throw new Error(`Policy compile error: ${message}`);
  }

  if (!result || typeof result !== 'object' || Array.isArray(result)) {
    throw new Error('Ejected policy code must return an object with policy overrides.');
  }

  const normalized = normalizePolicy({ ...base, ...(result as Partial<SandboxPolicy>) });
  const validationErrors = validatePolicy(normalized);
  if (validationErrors.length) {
    throw new Error(`Ejected policy invalid: ${validationErrors.join(' ')}`);
  }
  return normalized;
}

export function resolvePolicy(editorState: PolicyEditorState): SandboxPolicy {
  const normalized = normalizePolicyEditorState(editorState);
  if (normalized.mode === 'eject') {
    return compilePolicyCode(normalized.executableCode, normalized.template);
  }
  const validationErrors = validatePolicy(normalized.template);
  if (validationErrors.length) {
    throw new Error(`Template policy invalid: ${validationErrors.join(' ')}`);
  }
  return normalizePolicy(normalized.template);
}
